rust/src/traits/samplers/bernoulli/mod.tex

```diff
@@ -0,0 +1,327 @@
+
+Next \texttt{sample\_bernoulli} directly computes the number of leading zeroes in the binary expansion of \texttt{prob} to be $\texttt{leading\_zeroes} = \texttt{T::EXPONENT\_BIAS} - 1 - \texttt{raw\_exponent(prob)}$, where \texttt{raw\_exponent(prob)} is the value stored in the $\ell$ bits of the exponent. This value is correct by the specification of a $(k,\ell)$-bit float.  \texttt{sample\_bernoulli} then matches on the value \texttt{first\_heads\_index} corresponding to $I \sim Geom(0.5)$ returned by the function \texttt{sample\_geometric\_buffer}: \\
+
+\noindent\textbf{Case 1} ($\texttt{first\_heads\_index} < \texttt{leading\_zeroes}$). \\
+\noindent For any $I < \texttt{leading\_zeroes}$, it follows that $a_I = 0$ and therefore we should return \texttt{false}. In this case, \texttt{sample\_bernoulli} returns \texttt{false}.\\
+
+\noindent\textbf{Case 2} ($\texttt{first\_heads\_index} == \texttt{leading\_zeroes}$). \\
+\noindent When $\texttt{first\_heads\_index} == \texttt{leading\_zeroes}$, it follows that $a_I$ indexes into the implicit bit by the specification of a $(k,\ell)$-bit float. For all $(k,\ell)$-bit normal floats, the implicit bit equals $1$ and therefore we should return \texttt{true}. In this case, \texttt{sample\_bernoulli} returns \texttt{true}.\\
+
+\noindent\textbf{Case 3} ($ \texttt{leading\_zeroes} < \texttt{first\_heads\_index}  <  \texttt{leading\_zeroes}  + \texttt{T::MANTISSA\_BITS}$). \\
+\noindent When $ \texttt{leading\_zeroes} < \texttt{first\_heads\_index}  <  \texttt{leading\_zeroes}  + \texttt{T::MANTISSA\_BITS}$, $a_I$ corresponds to one of the bits in the mantissa. 
+\texttt{sample\_bernoulli}  left-shifts \texttt{1} by  $(\texttt{MANTISSA\_BITS + leading\_zeroes - first\_heads\_index})$ digits, the position in the binary representation of \texttt{prob} corresponding to $a_I$. Since the operation between the left-shifted \texttt{1} and the binary representation of \texttt{prob} at that position is a bitwise AND, if the bit in question is 1 (matching the left-shifted \texttt{1}), \texttt{sample\_bernoulli} will return 1. Otherwise, \texttt{sample\_bernoulli} will return 0. \\
```

Comment:
I'd be a bit more explicit here and in the pseudocode to check if the value is zero or not. I know you get implicit truthiness in python, but we really want to be returning a bool from this function, not an int.